docs/wiki/99 - Developers/Automated Testing.md

```diff
@@ -0,0 +1,91 @@
+
+Generally there are two categories of testing, **static** and **dynamic**. The following section breaks down the differences between the two and which tools are used for each.
+
+## Static Tests
+
+**Static tests** analyze the code without running it, checking for consistency between documentation and the source code.
+
+Static tests can be executed by running `make checkstatic` in the LORIS root diretory. This command will also use PHP's default linter (`php -l`) to check for basic syntax errors.
+
+We use the following tools for static testing.
+
+### PHP
+
+#### Phan
+
+> Phan looks for common issues and will verify type compatibility on various operations when type information is available or can be deduced.
+
+We use phan extensively to make sure that PHP code is taking advantage of types to produce stable results.
+
+Phan is configured in LORIS to use as many rules as possible. There are a list of excluded rules which are not evaluated on the codebase. Our goal is to eventually use every phan rule and gradually increase the strictness, and thus stability, of the codebase.
+
+The configuration file for phan, which includes a list of all excluded rules, can be found at `.phan/config.php`.
+
+For information about a specific phan rule, read their [breakdown of all issues types caught by phan](https://github.com/phan/phan/wiki/Issue-Types-Caught-by-Phan).
+
+#### PHP Code Sniffer (PHPCS)
+
+> PHP_CodeSniffer is a set of two PHP scripts; the main phpcs script that tokenizes PHP, JavaScript and CSS files to detect violations of a defined coding standard, and a second phpcbf script to automatically correct coding standard violations. PHP_CodeSniffer is an essential development tool that ensures your code remains clean and consistent.
+
+We use PHPCS to create a consistent coding style for PHP files. It also ensures that documentation is provided for functions and classes.
+
+A list of rules used by LORIS can be found in the file `test/LorisCS.xml`. These rules apply to the majority of the codebase. The `src/` folder in LORIS is configured to use different PHPCS rules that conform to the PSR2 standard. These rules can be found in `test/SrcCS.xml`.
+
+As of this writing, the authors of PHPCS have not created a document that explains each and every rule used by PHPCS. If you have issues with your code and PHPCS, please contact us and we'll try to help.
+
+PHPCS documentation can be found on [their GitHub wiki](https://github.com/squizlabs/PHP_CodeSniffer/wiki).
+
+
+#### PHP Mess Detector (PHPMD)
+
+Currently PHPMD is used by LORIS to check for unused code, e.g. variables that are declared but never use.
+
+Similar to `phan`, we aim to gradually increase the rules used by PHPMD to make LORIS clean and stable.
```

Comment:
I don't think our goal is to increase dependency on phpmd, it's a hard to use and understand tool and a lot of the rules are opinionated, while phan only identifies things that are bugs, not opinions.